### PR TITLE
ON-43678 # added support for invisible recaptcha

### PR DESCRIPTION
**Pull request by divporter**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [x] Have you tested your implementation locally?
- [x] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [ ] "Debugging" code removed
- [ ] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [ ] New files, variables and functions are descriptive/comprehensible and named consistently.
- [ ] There is no dead code (unreachable code)
- [ ] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [ ] There is no commented out code.
- [x] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

### CODE REVIEW

File: src/OneBlinkFormBase.tsx [hunk 1]
```diff
@@ -139,6 +141,11 @@ export type OneBlinkFormBaseProps = OneBlinkReadOnlyFormProps & {
    * for the structure of the argument and a sample function to be used.
    */
   onUploadAttachment?: typeof attachmentsService.uploadAttachment
+  /**
+   * Determines whether to use checkboxes or invisible recaptcha v2 for captcha
```
Comment (mymattcarroll) [inline src/OneBlinkFormBase.tsx:145]: Can we please add a something to this doc that indicates the default if leave as `undefined`

File: src/OneBlinkFormBase.tsx [hunk 2]
```diff
@@ -532,21 +550,33 @@ function OneBlinkFormBase({
     [definition],
   )
 
+  const addCaptchaRef = React.useCallback(
```
Comment (mymattcarroll) [inline src/OneBlinkFormBase.tsx:553]: This function should return a function to deregister the `ReCAPTCHA` so that the `useEffect()`s further down the tree can remove them if they get unmounted for any reason.

File: src/OneBlinkFormBase.tsx [hunk 3]
```diff
@@ -186,9 +193,13 @@ function OneBlinkFormBase({
   taskGroup,
   taskGroupInstance,
   onUploadAttachment,
+  captchaType,
 }: Props) {
   const isOffline = useIsOffline()
   const { isUsingFormsKey, userProfile } = useAuth()
+  const [captchaRefs, setCaptchaRefs] = React.useState<
+    Array<React.RefObject<ReCAPTCHA>>
```
Comment (mymattcarroll) [inline src/OneBlinkFormBase.tsx:201]:
```suggestion
    ReCAPTCHA[]
```

No need to register a ref object, better off just registering a  `ReCAPTCHA` object.

File: src/OneBlinkFormBase.tsx [hunk 4]
```diff
@@ -994,8 +1055,13 @@ function OneBlinkFormBase({
...
                           >
-                            <CaptchaSiteKeyContext.Provider
-                              value={captchaSiteKey}
+                            <CaptchaContext.Provider
+                              value={{
+                                captchaSiteKey,
+                                captchaType: recaptchaType,
+                                captchaRefs,
+                                addCaptchaRef,
+                              }}
```
Comment (mymattcarroll) [inline src/OneBlinkFormBase.tsx:1064]: This value should be memo-ised to avoid unnecessary re-renders

File: src/hooks/useCaptcha.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { useContext, createContext, RefObject } from 'react'
+import { ReCAPTCHA } from 'react-google-recaptcha'
+import { CaptchaType } from '../types/form'
+
+export const CaptchaContext = createContext<{
+  captchaSiteKey: string | undefined
+  captchaType: CaptchaType
+  captchaRefs: Array<RefObject<ReCAPTCHA>>
```
Comment (mymattcarroll) [inline src/hooks/useCaptcha.ts:8]: is this needed?

File: src/OneBlinkFormBase.tsx [hunk 6]
```diff
@@ -532,21 +548,45 @@ function OneBlinkFormBase({
     [definition],
   )
 
+  const addCaptchaRef = React.useCallback((recaptcha: ReCAPTCHA) => {
+    captchasRef.current.push(recaptcha)
+    // this allows the FormElementCaptcha element to unregister any captchas
+    return () => {
+      captchasRef.current = captchasRef.current.filter(
+        (recaptchaInstance) => recaptchaInstance === recaptcha,
```
Comment (mymattcarroll) [inline src/OneBlinkFormBase.tsx:556]:
```suggestion
        (recaptchaInstance) => recaptchaInstance !== recaptcha,
```

File: src/OneBlinkFormBase.tsx [hunk 7]
```diff
@@ -532,21 +548,45 @@ function OneBlinkFormBase({
...
+    })
     setHasAttemptedSubmit(false)
     setFormSubmission((current) => {
       return { ...current, submission: updatedModel }
     })
   }, [definition.elements, setFormSubmission, submission])
 
   const handleSubmit = React.useCallback(
-    (
+    async (
```
Comment (mymattcarroll) [inline src/OneBlinkFormBase.tsx:589]: If we are making this async, we need to: 
1. wrap in a try/catch to handle errors and also show some 
2. Disable the submit button while processing to prevent double clicking
3. show something to indicate something is happening in case the token(s) takes a long time to generate e.g. loading animation

Comment (divporter) [reply src/OneBlinkFormBase.tsx:591]: All settled should handle 1
